src/psyclone/psyir/backend/fortran.py

```diff
@@ -947,7 +1010,9 @@ def gen_decls(self, symbol_table, is_module_scope=False):
         unresolved_symbols = []
         for sym in all_symbols[:]:
             if isinstance(sym.interface, UnresolvedInterface):
-                unresolved_symbols.append(sym)
+                # Explicitly deal with '*' as in 'class(*) :: var'
+                if sym.name != "*":
```

Comment:
Indeed, but that's how the frontend deals with `whatever` in either `class(whatever)` or `type(whatever)`. I'm not so sure how to deal with this case unless I modify the frontend even more extensively for this (awful) Fortran syntax? It's used in tests but I certainly hope no one actually uses `class(*)` polymorphism... at least it's not in any `.f90` test file.
I've moved dealing with this '*' symbol in the backend (by removing it so that it can't be declared anywhere) to line 1003, as this feels a bit cleaner.
Let me know what you think

Reply:
I think this means we have to modify the frontend a bit more - conceptually a Symbol refers to a memory location. Here, the "*" is syntax meaning (I've just googled) "unlimited polymorphic type" apparently. I've a feeling this is used in LFRic and I'm sure other codes will as well. So, when we encounter a declaration like:
```fortran
  class(*) :: some_arg
```
we know we have a Symbol named "some_arg" but we don't know the type. The simplest thing to do would be to just raise a NotImplementedError in this case so that we will get a Symbol of UnsupportedFortranType. If that's not sufficient for your particular use case, we maybe need a new type. We already have `UnresolvedType` - we could perhaps add an `is_class` property to that? What do @sergisiso and @hiker think?

Reply:
I feel like using `UnsupportedFortranType` would be going slightly backward here, as we'd be back to string declarations for these symbols (and I'd have to edit a lot of tests again due to upper/lowercase fparser/PSyclone outputs, probably).

I'm open to whatever other option, if having this '*' symbol is too problematic.